Convert to python using robocorp libraries

OvidiuCode opened the pull request:
Convert robot to use robocorp python libraries. 

Known caveats:

1. Some pages have lazy image load so the screenshot right now will only contain content that was loaded on the page. To fix this a scroll to the bottom of the page can be implemented.
2. In some cases the cookie accept or data consent has an animation and if we do the screenshot too early we might capture that pop-up even if we accepted the cookie. For this we implemented the a wait of 1 second. A better approach would be to check if the pop-up does not exist anymore.

OvidiuCode committed:
Convert to python using robocorp libraries

cmin764 commented on README.md:
```diff
@@ -2,25 +2,19 @@
 [![YouTube video](https://img.youtube.com/vi/aQkXTHP3Xxw/0.jpg)](http://www.youtube.com/watch?v=aQkXTHP3Xxw)
 
-- Reads the website data (URL, accept cookies selector) from a `.csv` file using the [`RPA.Tables`](https://robocorp.com/docs/libraries/rpa-framework/rpa-tables) library.
+- Reads the website data (URL, accept cookies and consent selectors) from a `.csv` using the python csv built in library.
```
Nit:
```suggestion
- Reads the website data (URL, accept cookies and consent selectors) from a `.csv` using the Python's `csv` built-in library.
```

cmin764 commented on conda.yaml:
🐛 Any reason for why we should use Py 3.10.**13**? And it looks like we've missed the necessary `truststore` dep to fix the SSL trouble on systems requiring SSL certs to access the network.

We can follow the browser template [example](https://github.com/robocorp/template-python-browser/blob/master/conda.yaml), where we should use the latest `robocorp-*` versions, but keeping Python and pip versions inline with the template to avoid stability issues in the customer machine.

Could be:
```yaml
channels:
  - conda-forge
dependencies:
  - python=3.10.12
  - robocorp-truststore=0.8.0
  - pip=23.2.1
  - pip:
      - robocorp==1.6.0
      - robocorp-browser==2.2.2
```

@vjmp Can tell you more about the pip stability and the importance of having these inline.

cmin764 commented on tasks.py:
```diff
@@ -0,0 +1,60 @@
+import time
+from urllib.parse import urlparse
+from robocorp.tasks import task
+from robocorp import browser
```
Nit: wouldn't hurt to have one space separating the stdlib from 3rd-party imports, then some sorting. (if Ruff/isort doesn't do that already in your editor)

cmin764 commented on tasks.py:
```diff
@@ -0,0 +1,60 @@
+@task
+def take_website_screenshots():
+    """Take screenshots of all websites found in websites.csv file."""
+    with open("websites.csv", mode="r") as csv_file:
```
Nit: made a practice to put dev data in a dir called `devdata` so you can move this file there if you like -> _devdata/websites.csv_

No strong-opinion, just an alignment nit and your decision.

cmin764 commented on tasks.py:
```diff
@@ -0,0 +1,60 @@
+
+        accept_cookies_and_consents(website)
+
+        # some websites have animations on cookies and consents so wait for that to disappear
+        time.sleep(1)
```
Rather than waiting (in case you're not sure this won't go past 1s), you may test the presence of the consent selector(s) in a loop (small timeout & sleep duration), and once they aren't visible anymore, you proceed with the screenshots.
```python
    # Use Playwright's query_selector method to check for the selector
    element = page.query_selector("your-selector")
    
    # Check if the element exists
    if element:
        print("Element exists.")
    else:
        print("Element does not exist.")
```

An alternative would be to have another selector in your _.csv_ file which tells you when the page is ready in case a new element would be visible at that moment. And then you wait for it with `page.wait_for_selector("your-selector")`.

----

But if is guaranteed that the sleep is long enough and the animation goes away quickly, then it doesn't pay the effort to check for page readiness.

OvidiuCode replied:
Makes sense, but considering only one page has this behaviour from the list, I think it is easier for people to follow with a time.sleep() rather than introducing another selector / another column in the csv.

cmin764 commented on tasks.py:
```diff
@@ -0,0 +1,60 @@
+        time.sleep(1)
+
+        domain = get_domain_name(website["url"])
+        page.screenshot(path=f"output/{domain}.png", full_page=True)
```
🐛 We usually define a constant for the output dir:
```python
OUTPUT_DIR = Path(os.environ.get("ROBOT_ARTIFACTS", "output"))
```
and then you can do:
```suggestion
        page.screenshot(path=str(OUTPUT_DIR / f"{domain}.png"), full_page=True)
```

cmin764 commented on tasks.py:
```diff
@@ -0,0 +1,60 @@
+
+def get_domain_name(url) -> str:
+    parsed_url = urlparse(url)
+    domain_parts = parsed_url.netloc.split(".")
```
Nit
```suggestion
    return parsed_url.netloc.replace(".", "_")
```
in case you become afraid of clashes like:
- site.com vs. site.ro
- app.site.com vs. news.site.com
where in each case a `site.png` image will be generated.

OvidiuCode replied:
The alternative here is to leave only the network location as it was in the robot.yaml example

cmin764 replied:
That doesn't sound bad, the more explicit the better, just don't forget to replace the dots (`.`) with something like underscores.

cmin764 commented on tasks.py:
```diff
@@ -0,0 +1,60 @@
+
+    if website["accept_cookies_selector"]:
+        try:
+            page.locator(website["accept_cookies_selector"]).click()
```
```python
ACCEPT_COOKIES_COLUMN = "accept_cookies_selector"  # taken from the header of the .csv input file
```

```suggestion
    selector = website.get(ACCEPT_COOKIES_COLUMN)
    if selector:
        try:
            page.locator(selector).click()
```

Might be a lot easier to understand and refactor next time we'd change the structure of the _.csv_ file.

----

The same can apply for the below block with the data consent.

cmin764 commented on tasks.py:
```diff
@@ -0,0 +1,60 @@
+        try:
+            page.locator(website["accept_cookies_selector"]).click()
+        except Exception:
+            pass
```
Would be good to explicitly display exceptions (if these aren't already reported during the exception handling itself -- need to check):
```suggestion
        except Exception as exc:
            log.warn(f"An error occurred during the cookie consent: {exc}")
```

cmin764 commented on websites.csv:
Awesome, here's a list of more publications if you ever need the inspiration to add [more](https://github.com/cmin764/TrueStory/blob/develop/truestory/data/media-bias.csv).

cmin764 reviewed (changes requested):
Looks good overall, good job! -- ⚠️ I haven't run the example yet!

Left some comments (like the output dir resolution) which would be nice to be aligned with the rest of the examples, and some nits which are purely optional and I don't have a strong opinion over them.

➕  Would be nice to have at the end of the README some reference links pointing you to the docs (how to identify web locators, the Inspector, the browser library, automation tutorial for Browser etc.) -- similar [example](https://github.com/robocorp/example-windows-notepad?tab=readme-ov-file#links) for Windows

----

Beside this, some grooming over the GH PR itself won't hurt, like:

1. Assigning who handles it (assigned you already).
2. Who should review + request for review (me).
3. Basic description of the PR itself: we know what is this about already from the title (conversion), but how about what it adds extra and what it changes in terms of behaviour, input sources, the way it make screenshots now if different and known caveats (if there are any), things which help the reviewer.

OvidiuCode committed:
Updates based on review

Commit:
Post-review fixes:

- Distinctive task name to detect and run in VSCode
- Docstrings following our standard
- Escaped screenshot names

cmin764 reviewed (approved):
This is in a stellar shape now, thank you for your work!

Also left a small [commit](https://github.com/robocorp/example-full-page-website-screenshots/pull/2/commits/befd6f9c6528e65d23a0eb85a3e4194096076249) from my side if you don't mind. (feel free to re-refactor if you want to improve that further)
